components/collections/codepointtrie_builder/cpp/list_to_ucptrie.cpp

```diff
@@ -2,111 +2,57 @@
-		x = x * 10 + (ch - '0');
-	}
-	return x * sgn;
-}
+#include "unicode/errorcode.h"
+#include "unicode/umutablecptrie.h"
+#include "writesrc.h"
 
 /**
  * list_to_ucptrie: converts a stream of integers to a UCPTrie
  *
- * - standard input should be a stream of integers, each integer representing the value for
-     consecutive code points, starting from code point 0.
+ * - standard input should be a stream of integers, each integer representing
+ the value for consecutive code points, starting from code point 0.
  * - standard output is the build UCPTrie as a TOML file
  */
-int main(int argc, char const *argv[]) {
-    if (argc != 5) {
-        std::cerr << "Takes 4 positional arguments: default value, error value, trie type, and value width" << std::endl;
-        return 1;
-    }
-
-    int32_t defaultValue = atoi(argv[1]);
-    int32_t errorValue = atoi(argv[2]);
-
-    UCPTrieType trieType;
-    std::string trieTypeStr(argv[3]);
-    if (trieTypeStr == "small") {
-        trieType = UCPTRIE_TYPE_SMALL;
-    } else if (trieTypeStr == "fast") {
-        trieType = UCPTRIE_TYPE_FAST;
-    } else {
-        std::cerr << "Expected 'small' or 'fast' for 3rd argument" << std::endl;
-        return 1;
-    }
 
-    UCPTrieValueWidth valueWidth;
-    int32_t valueWidthInt = atoi(argv[4]);
-    if (valueWidthInt == 8) {
-        valueWidth = UCPTRIE_VALUE_BITS_8;
-    } else if (valueWidthInt == 16) {
-        valueWidth = UCPTRIE_VALUE_BITS_16;
-    } else if (valueWidthInt == 32) {
-        valueWidth = UCPTRIE_VALUE_BITS_32;
-    } else {
-        std::cerr << "Expected '8', '16', or '32' for 4th argument" << std::endl;
-        return 1;
-    }
-
-    icu::ErrorCode status;
-
-    icu::LocalUMutableCPTriePointer builder(umutablecptrie_open(
-        defaultValue,
-        errorValue,
-        status));
-    if (status.isFailure()) {
-        std::cerr << status.errorName() << std::endl;
-        return 1;
-    }
-
-    UChar32 cp = 0;
-    int32_t n = read_int32_t();
-    int32_t value = 0;
-    while(n--) {
-        value = read_int32_t();
-        umutablecptrie_set(builder.getAlias(), cp, value, status);
-        if (status.isFailure()) {
-            std::cerr << status.errorName() << std::endl;
-            return 1;
-        }
-        cp++;
-    }
-    std::cerr << "Saved values for code points up to " << cp << std::endl;
-
-    icu::LocalUCPTriePointer utrie(umutablecptrie_buildImmutable(
-        builder.getAlias(),
-        trieType,
-        valueWidth,
-        status));
+extern "C" {
+int construct_ucptrie(const int32_t defaultValue, const int32_t errorValue,
+                      const UCPTrieType trieType,
+                      const UCPTrieValueWidth valueWidth,
+                      const uint32_t* values, const uint32_t n) {
+  icu::ErrorCode status;
+
+  icu::LocalUMutableCPTriePointer builder = icu::LocalUMutableCPTriePointer(
+      umutablecptrie_open(defaultValue, errorValue, status));
```

Comment:
This line has implicit casts: `defaultValue` and `errorValue` are `int32_t` that are getting cast to `uint32_t`. I noticed this because the wasm signature of `construct_ucptrie` is `<(i32, i32, i32, i32, i32, i32), i32>` and you have to cast everything to `i32`, but there really shouldn't be any unsigned values anywhere. Please fix.